subsys/dfu/boot/mcuboot.c

```diff
@@ -0,0 +1,254 @@
+#define BOOT_FLAG_UNSET 0xff
+/* end_of Strict defines */
+
+#define BOOT_MAGIC_GOOD  1
+#define BOOT_MAGIC_BAD   2
+#define BOOT_MAGIC_UNSET 3
+
+#define BOOT_FLAG_IMAGE_OK 0
+#define BOOT_FLAG_COPY_DONE 1
+
+#define COPY_DONE_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE -\
+				   BOOT_MAGIC_SZ - BOOT_MAX_ALIGN * 2)
+
+#define IMAGE_OK_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE - BOOT_MAGIC_SZ -\
+				  BOOT_MAX_ALIGN)
+#define MAGIC_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE - BOOT_MAGIC_SZ)
+
+const u32_t boot_img_magic[4] = {
+	0xf395c277,
+	0x7fefd260,
+	0x0f505235,
+	0x8079b62c,
+};
+
+static struct device *flash_dev;
+
+static int boot_flag_offs(int flag, u32_t bank_offs, u32_t *offs)
+{
+	switch (flag) {
+	case BOOT_FLAG_COPY_DONE:
+		*offs = COPY_DONE_OFFS(bank_offs);
+		return 0;
+	case BOOT_FLAG_IMAGE_OK:
+		*offs = IMAGE_OK_OFFS(bank_offs);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int boot_write_flag(int flag, u32_t bank_offs)
+{
+	u8_t buf[FLASH_MIN_WRITE_SIZE];
+	u32_t offs;
+	int rc;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	memset(buf, 0xFF, sizeof(buf));
+	buf[0] = BOOT_FLAG_SET;
+
+	flash_write_protection_set(flash_dev, false);
+	rc = flash_write(flash_dev, offs, buf, sizeof(buf));
+	flash_write_protection_set(flash_dev, true);
+
+	return rc;
+}
+
+static int boot_flag_get(int flag, u32_t bank_offs)
+{
+	u32_t offs;
+	int rc;
+	u8_t flag_val;
+
+	rc = boot_flag_offs(flag, bank_offs, &offs);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = flash_read(flash_dev, offs, &flag_val, sizeof(flag_val));
+	if (rc != 0) {
+		return rc;
+	}
+
+	return flag_val;
+}
+
+static u8_t boot_copy_done_get(u32_t bank_offs)
+{
+	return boot_flag_get(BOOT_FLAG_COPY_DONE, bank_offs);
+}
+
+static int boot_image_ok_get(u32_t bank_offs)
+{
+	return boot_flag_get(BOOT_FLAG_IMAGE_OK, bank_offs);
+}
+
+static int boot_write_image_ok(u32_t bank_offs)
+{
+	return boot_write_flag(BOOT_FLAG_IMAGE_OK, bank_offs);
+}
+
+static int boot_write_magic(u32_t bank_offs)
+{
+	u32_t offs;
+	int rc;
+
+	offs = MAGIC_OFFS(bank_offs);
+
+	flash_write_protection_set(flash_dev, false);
+	rc = flash_write(flash_dev, offs, boot_img_magic, BOOT_MAGIC_SZ);
+	flash_write_protection_set(flash_dev, true);
+
+	return rc;
+}
+
+static int boot_magic_code_check(const u32_t *magic)
+{
+	int i;
+
+	if (memcmp(magic, boot_img_magic, sizeof(boot_img_magic)) == 0) {
+		return BOOT_MAGIC_GOOD;
+	}
+
+	for (i = 0; i < 4; i++) {
+		if (magic[i] == 0xffffffff) {
+			return BOOT_MAGIC_UNSET;
+		}
+	}
+
+	return BOOT_MAGIC_BAD;
+}
+
+static int boot_magic_state_get(u32_t bank_offs)
+{
+	u32_t magic[4];
+	u32_t offs;
+	int rc;
+
+	offs = MAGIC_OFFS(bank_offs);
+	rc = flash_read(flash_dev, offs, magic, sizeof(magic));
+	if (rc != 0) {
+		return rc;
+	}
+
+	return boot_magic_code_check(magic);
+}
+
+int boot_img_pending_set(int permanent)
+{
+	int rc;
+
+	switch (boot_magic_state_get(FLASH_BANK1_OFFSET)) {
```

Comment:
I don't quite understand the logic, or for that matter why there is so much code here.  Honestly, the only thing all of this code needs to do is, after placing the image in place, write the magic at the end, and write the "image_ok" marker at the end. There doesn't need to be any logic, or even reading of the device, since the partition should be erased by the process of writing the image.